bundles-lib/src/main/java/org/apache/metron/bundles/ExtensionManager.java

```diff
@@ -0,0 +1,436 @@
+      // Must set the context class loader to the bundle classloader itself
+      // so that static initialization techniques that depend on the context class loader will work properly
+      final ClassLoader ncl = bundle.getClassLoader();
+      Thread.currentThread().setContextClassLoader(ncl);
+      loadExtensions(bundle);
+
+      // Create a look-up from withCoordinates to bundle
+      bundleCoordinateBundleLookup.put(bundle.getBundleDetails().getCoordinates(), bundle);
+    }
+
+    // restore the current context class loader if appropriate
+    if (currentContextClassLoader != null) {
+      Thread.currentThread().setContextClassLoader(currentContextClassLoader);
+    }
+  }
+
+  /**
+   * Returns a bundle representing the system class loader.
+   *
+   * @param bundleProperties a BundleProperties instance which will be used to obtain the default
+   * Bundle library path, which will become the working directory of the returned bundle
+   * @return a bundle for the system class loader
+   */
+  public static Bundle createSystemBundle(final FileSystemManager fileSystemManager,
+      final BundleProperties bundleProperties) throws FileSystemException, URISyntaxException {
+    final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+
+    final String bundlesLibraryDirectory = bundleProperties
+        .getProperty(BundleProperties.BUNDLE_LIBRARY_DIRECTORY);
+    if (StringUtils.isBlank(bundlesLibraryDirectory)) {
+      throw new IllegalStateException(
+          "Unable to create system bundle because " + BundleProperties.BUNDLE_LIBRARY_DIRECTORY
+              + " was null or empty");
+    }
+    final URI bundlesLibraryDirURI = bundleProperties.getBundleLibraryDirectory();
+    FileObject bundleDir = fileSystemManager.resolveFile(bundlesLibraryDirURI);
+
+    // Test if the source Bundles can be read
+    FileUtils.ensureDirectoryExistAndCanRead(bundleDir);
+
+    final BundleDetails systemBundleDetails = new BundleDetails.Builder()
+        .withBundleFile(bundleDir)
+        .withCoordinates(SYSTEM_BUNDLE_COORDINATE)
+        .build();
+
+    return new Bundle(systemBundleDetails, systemClassLoader);
+  }
+
+  /**
+   * Loads extensions from the specified bundle.
+   *
+   * @param bundle from which to load extensions
+   */
+  @SuppressWarnings("unchecked")
+  private static void loadExtensions(final Bundle bundle) throws NotInitializedException {
+    checkInitialized();
+    for (final Map.Entry<Class, Set<Class>> entry : definitionMap.entrySet()) {
+      // this is another extention point
+      // what we care about here is getting the right classes from the classloader for the bundle
+      // this *could* be as a 'service' itself with different implementations
+      // The NAR system uses the ServiceLoader, but this chokes on abstract classes, because for some
+      // reason it feels compelled to instantiate the class,
+      // which there may be in the system.
+      // Changed to ClassIndex
```

Comment:
I think the ClassIndex is a big improvement over the ServiceLoader too.  Easier to use, faster, doesn't try to instantiate abstracts.... safer because of the compile time generation of the service file....  just great.

I wonder if Nifi would take a PR with VFS / VFSClassloader / ClassIndex get
* better generation on the archetype side
* support for hdfs and other filesystems
* not having to manage the working directories etc
I'll have to talk to @bbende if we ever get done ;)